### ISSUE TITLE
All not implemented methods should raise exceptions

### ISSUE THREAD

**Issue by githubnemo** [SQLDroid/SQLDroid#32]
I don't think it is a good idea to pretend that not implemented methods work and simply return a dummy value, e.g. `null`.

Methods like 

```
@Override
public CallableStatement prepareCall(String sql) throws SQLException {
    System.err.println(" ********************* not implemented @ " + DebugPrinter.getFileName() + " line " + DebugPrinter.getLineNumber());

    return null;
}
```

should be converted to 

```
@Override
public CallableStatement prepareCall(String sql) throws SQLException {
    throw new UnsupportedOperationException("Not implemented yet");
}
```

Reasoning: it's way easier to debug. I just tracked down a `NullPointerException` because a prepared statement was `null` altough it should not have been. An exception about the method not being implemented would have helped.

**Comment by jhannes** [comment]
+1

I can implement this change if desired.

**Comment by donv** [comment]
@jhannes Go ahead! 👍

**Comment by ergouser** [comment]
Raising exceptions for most unimplemented methods is a good plan, however I think that there might be a few that are harmless - maybe status methods that return something useless rather than a valid status??

That is, I think that I occasionally see "not implemented" flash by but they don't seem to have any effect on the application (and, of course, I don't have any examples to hand).

Should be obvious if such cases exist, but it does make these changes more than a find/replace or sed substitution.

**Comment by jhannes** [comment]
@ergouser I've worked my way through ResultSet. There are a number of methods in ResultSet where the spec allows for SQLFeatureNotImplementedException, and I've used that in these cases. I've left a few  warnings, for example for getTimestamp(int, Calendar).

Luckily all the updateXXX could be corrected with some editor shortcuts. :-)

Feel free to chip in on methods that should behave differently.

**Comment by ergouser** [comment]
@jhannes Looks good.  I was thinking of methods like that and setFetchSize where ignoring the failure is (mostly) harmless.

**Comment by jhannes** [comment]
I have added implementation or exception for pretty much everything now. I've added TODOs for all remaining System.err lines. I think the issue can be closed, but I'd like a review from @donv, @ergouser and @githubnemo first.

To review:
* TODOs saying "Evaluate if implementation is sufficient" - delete the comment and System.err if you think so
* All 67 TODOs if there's anything especially strange
* "throw new UnsupportedOperationException" (I have not updated SQLDroidClob, so there are 30 places where I added this). Are there any places where a noop would be better?

**Comment by ergouser** [comment]
I don't have time to test this or build the tests, but I just hate to see the amount of effort expended in that method and then it just throws...

Blob:   	public byte[] getBytes(long pos, int length) throws SQLException {

add:

  if (pos >= b.length ) {
 +      throw new SQLException("pos must be < blob length");
 +    }
 +    if (length > b.length ) {
 +      throw new SQLException("length must be < blob length");
 +    }

Replace:

      throw new UnsupportedOperationException("getBytes implementation only supports return full byte array");

>>> 
//not horribly efficient for a very large blob, but it should return a value
      byte[] tmp = new byte[length];
      System.arraycopy(b, pos, tmp, 0, length);
return tmp;

>>

**Comment by jhannes** [comment]
@ergouser Good idea, although it's not really part of this bug. :-) I've added assertions to a test and your implementation.

System.arraycopy may be slow for large values, but there is no other way to create a new byte array, so this has to be the implementation

**Comment by ergouser** [comment]
There seem to be a few methods in Blob that would not be difficult to implement, but probably best left for another time.  Unrelated: there's enough activity that I'm starting to feel guilty about neglecting the non-Android portion.  The version I'm using works well, indeed I use the code as much off-Android as on, but it's been a while since I updated my installation.  I seem to be the only person who really cares about desktop platforms, but in fact SQLDroid is about the best JDBC driver for SQLite on any platform.

**Comment by jhannes** [comment]
@ergouser Let's leave the Blob implementation for another issue.

I think most of the activity is my frequent commits, so I wouldn't feel too guilty if I were you. :-)

However, I think you just blew my mind: Do you use SQLDroid on non-Android Java implementations as well? That's awesome! How do you do it?

**Comment by ergouser** [comment]
It uses JNA and the native code provided by SQLite.

Basically, build a jar, say "SQLDroidCompat.jar" from:
compat/android
add that, the jna.jar and the appropriate .so, .dll, etc.for you platform from the SQLite website to your project (as well as the SQLDroid jar).

I've used it on Windows (32 & 64) and Linux. We generate code that works on pretty much any platform, including Android.  Any JDBC driver _should_ work, but using the same driver ensures compatibility.

**Comment by donv** [comment]
This is a great feature and should be documented and linked from the README.  Can you do that @ergouser ?  Any format will do, but I think I would prefer a Markdown (.md) file in a `doc` folder.

Build instruction, a simple example, and a link to where one can get the `jna.jar` should do it.

**Comment by ergouser** [comment]
I've created a new issue (#73) to track this.  I won't have time for it for a little while, but I'll try to get back to it whenever I can.
